core/templates/dev/head/pages/library/SearchBarDirective.js

```diff
@@ -43,6 +43,68 @@ oppia.directive('searchBar', [
               }
             )
           );
+          $scope.activeMenuName = '';
+          $scope.ACTION_OPEN = 'open';
+          $scope.ACTION_CLOSE = 'close';
+          $scope.KEYBOARD_EVENT_TO_KEY_CODES = {
+            enter: {
+              shiftKeyIsPressed: false,
+              keyCode: 13
+            },
+            tab: {
+              shiftKeyIsPressed: false,
+              keyCode: 9
+            },
+            shiftTab: {
+              shiftKeyIsPressed: true,
+              keyCode: 9
+            }
+          };
+          /**
+           * Opens the submenu.
+           * @param {object} evt
+           * @param {String} menuName - name of menu, on which
+           * open/close action to be performed (category,language).
+           */
+          $scope.openSubmenu = function(evt, menuName) {
+            // Focus on the current target before opening its submenu.
+            angular.element(evt.currentTarget).focus();
+            $scope.activeMenuName = menuName;
+          };
+          $scope.closeSubmenu = function(evt) {
+            $scope.activeMenuName = '';
+            angular.element(evt.currentTarget).closest('li')
+              .find('a').blur();
+          };
+          /**
+           * Handles keydown events on menus.
+           * @param {object} evt
+           * @param {String} menuName - name of menu to perform action
+           * on(category/language)
+           * @param {object} eventsTobeHandled - Map keyboard events('Enter') to
+           * corresponding actions to be performed(open/close).
+           *
+           * @example
+           *  onMenuKeypress($event, 'category', {enter: 'open'})
+           */
+          $scope.onMenuKeypress = function(evt, menuName, eventsTobeHandled) {
```

Comment:
These are duplicate codes, isn't it? Can we find a way to remove such duplicate codes (maybe through a service?)

Reply:
found duplicates [here](https://github.com/oppia/oppia/blob/develop/core/templates/dev/head/components/top_navigation_bar/TopNavigationBarDirective.js#L163).

Reply:
@DubeySandeep can we file a different issue for this?

Reply:
Nope (why not in this PR ?)

Let us know if you are stuck anywhere.